soup/man/man1/location.1
.\" section 1 man page for location
.\"
.\" This man page was first written by Landon Curt Noll for the IOCCC
.\" in 2023 with minor improvements and fixes by Cody Boone Ferguson.
.\"
.\" Location is relative. :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH location 1 "20 February 2025" "location" "IOCCC tools"
.SH NAME
.B location
\- look up ISO 3166 codes, location names by name (substring, full) or print the entire table
.SH SYNOPSIS
.B location
.RB [\| \-h \|]
.RB [\| \-v
.IR level \|]
.RB [\| \-V \|]
.RB [\| \-c \|]
.RB [\| \-n \|]
.RB [\| \-s \|]
.RB [\| \-a \|]
.RI [\| location
.IR ... \|]
.SH DESCRIPTION
.B location
without any arguments will print the entire table of ISO 3166 codes, followed by a tab, followed by the formal location name for the given ISO 3166 code.
.sp 1
Given arguments,
.B location
tool will assume the arguments are ISO 3166 code(s) and print corresponding formal location name(s).
.sp 1
Arguments are case insensitive.
.sp 1
ISO 3166 code(s) are Alpha-2 codes: two UPPER case characters.
.sp 1
With
.BR \-n ,
the arguments
are assumed to be the formal location name(s).
The
.B location
tool will print their corresponding ISO 3166 code(s).
.PP
To search by substring use the
.B \-s
option.
To show all matches use the
.B \-a
option.
Showing all matches is less useful without the
.B \-n
option because country codes are two characters and are exact matches and the option is only relevant if an arg is specified.
The use of
.B \-s
requires an arg to the program much like
.B \-n
but it does not require the
.B \-n
option itself.
.PP
If verbosity is > level 0 and
.B \-a
is used it will show a string that shows name and code to more easily identify each match.
.SH OPTIONS
.TP
.B \-h
Show help and exit.
.TP
.BI \-v\  level
Set verbosity level to
.IR level
(def: 0).
.TP
.B \-V
Show version and exit.
.TP
.B \-c
Use common name, otherwise by default, the official name is used.
.TP
.B \-n
Assume the arguments are formal location name(s) and print
their corresponding ISO 3166 code(s).
.sp
By default the arguments are ISO 3166 code(s)
and their corresponding formal location name(s) are printed.
.sp
The use of
.B \-n
requires at least one argument.
.TP
.B \-s
Search by substring instead of by an exact match.
.sp
Use of
.B \-s
requires at least one arg.
.TP
.B \-a
Show all matches.
.sp
Use of
.B \-a
is less useful without the
.B \-n
option and only checked for if at least one arg is specified but nonetheless it is possible to use
.B \-a
without
.BR \-n .
.SH EXIT STATUS
.TP
0
All is OK
.TQ
1
At least one location argument could not be found
.TQ
2
.B \-h
and help string printed or
.B \-V
and version string printed
.TQ
3
command line error
.TQ
>= 4
internal error
.SH BUGS
.PP
If you have an issue with the tool you can report it at
.br
.IR \<https://github.com/ioccc\-src/mkiocccentry/issues\> .
.SH EXAMPLES
.PP
Print the entire location table:
.sp
.RS
.ft B
 location
.ft R
.RE
.sp
The table is printed as a two character ISO 3166 code
(in UPPER CASE), followed by a tab, followed by
the corresponding formal location name.
.PP
Print the formal location name given ISO 3166 codes:
.sp
.RS
.B location AU CA UA US XX
.RE
.PP
The ISO 3166 codes may be in any case:
.sp
.RS
.B location CX AU cA cz gb us Ie
.RE
.sp
.PP
Print the ISO 3166 codes given a formal location name:
.sp
.RS
.B location 'Equatorial Guinea'
.RE
.PP
Show all matches with the substring
.I united
in the name:
.sp
.RS
.B location \-a \-s \-n united
.RE
.sp
.PP
Show all matches with the substring
.I germ
in the name, showing both name and code to more easily identify each match:
.sp
.RS
.B location \-a \-s \-n \-v 1 germ
.RE
.sp
We note that, as the above example shows, Germany is a country full of 'germ ridden people' but we do not mean offence by this: it's just a fun pun of one of the author's.
.PP
Search by substring for both Unit and Germ with a verbosity of level 1, to show both name and code:
.sp
.RS
.B location \-asnv1 Unit Germ
.RE
.sp
This should show:
.RS
United Arab Emirates ==> AE
.br
United Kingdom of Great Britain and Northern Ireland (the) ==> GB
.br
United States Miscellaneous Pacific Islands ==> PU
.br
Tanzania, United Republic of ==> TZ
.br
United Kingdom ==> UK
.br
United States Minor Outlying Islands ==> UM
.br
United Nations ==> UN
.br
United States of America ==> US
.br
German Democratic Republic ==> DD
.br
Germany ==> DE
.br
.RE
.SH EDITORIAL NOTES
.sp
The primary source of the location table is:
.sp
.in +0.5i
.IR \<https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Decoding_table\> .
.in -0.5i
.sp
Additional information may be found at:
.sp
.in +0.5i
.IR \<https://en.wikipedia.org/wiki/ISO_3166-1#Officially_assigned_code_elements\>
.br
.IR \<https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2\>
.br
.IR \<https://www.iso.org/obp/ui/#iso:pub:PUB500001:en\>
.br
.IR \<https://www.iso.org/glossary-for-iso-3166.html\>
.br
.IR \<https://www.iso.org/obp/ui/#search\>
.in -0.5i
.sp
The location table was initially compiled on 2023 June 04
as follows:
We added all Officially assigned code elements,
all Exceptionally reserved code elements,
all Transitionally reserved code elements,
all Indeterminately reserved code elements,
all Formerly assigned code elements,
and all User-assigned code elements.
.sp
This means you will find a few codes that belong to
entities that no longer are recognized, entities
that are/were temporary, codes that are reserved for
future use, codes for Road Traffic Conventions,
codes for Property Organizations, duplicate codes, etc.
.sp
We do not list WIPO codes as they are not formally reserved.
They refer to Patent and related WIPO
organizations: as such they do not represent places
where an IOCCC winner would live.
.sp
In a number of cases, the common name is used in
place of a longer formal location name.
.sp
Because
.BR mkiocccentry (1)

CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 2.3.44 2025-02-26

Remove unused arg in `write_json_files()`.

Updated `MKIOCCCENTRY_VERSION` to `"1.2.34 2025-02-26"`.


## Release 2.3.43 2025-02-25

Improve `mkiocccentry_test.sh` - add only required files test.

Updated `MKIOCCCENTRY_TEST_VERSION` to `"1.0.14 2025-02-25"`.

Plug a loophole in `chkentry(1)` and make it do even more tests that were
missing/not previously thought of.

Updated `CHKENTRY_VERSION` to `"1.1.5 2025-02-25"`.


## Release 2.3.42 2025-02-24

Disable `-g3` in Makefile as use of `-O` option (besides level 0) makes
debugging much harder so one should not typically use both. This was done in
`dbg` and `dyn_array` (and opened a pull request in both repos) too; I did
`jparse/` (in the repo) a while back for the same reason as cited here.

Fun facts: the repo version is: `"2.3.42 2025-02-24"` and so the last two digits
of the version number itself and the last two digits of the date in the version
are opposites (and thus anagrams) for [42, the Ultimate Answer to/of Life, the
Universe and Everything](https://hitchhikers.fandom.com/wiki/42); but no matter
how absurd it is, and no matter how obvious it was a joke and no matter how much
[Douglas Adams explicitly said it was a
joke](https://groups.google.com/g/alt.fan.douglas-adams/c/595nPukE-Jo/m/koaAJ3tPBtEJ?pli=1),
people still think it's serious and people still have stated it is real and some
[astronomers even spent years studying
it](https://www.the-independent.com/news/yes-the-answer-to-the-universe-really-is-42-1351201.html)
only to claim that it **really is** 42! That means that the **real**
[Vogons](https://hitchhikers.fandom.com/wiki/Vogon) are
[humans](https://en.wikipedia.org/wiki/Human), at least until that [hyperspace
bypass](https://hitchhikers.fandom.com/wiki/Bypass) is formed. :-)

BTW: Landon told me that: at Google Headquarters (previously SGI) at building 42
there is a very hidden copy of The Hitchhiker's Guide to the Galaxy (H2G2) with
drawings and other things that Douglas Adams requested Landon taunt them
[Google] to find and he's had a lot of joy in doing so since very few have been
able to find it (and Landon is one of the very few who know about it).

Okay so there's some fun totally unrelated to the changes but it's fun so why
not? :-)

Fix bug in `mkiocccentry` where a submission topdir which only had the three
required files would abort with an error (due to list being not NULL but empty
due to a typo).

Sync [jparse repo](https://github.com/xexyl/jparse/) to `jparse/` for some
upcoming fixes to `chkentry` (I thought of a loophole in a new option that was
being finalised and more was discovered after this).

`chkentry` has had a partial fix but more needs to be done (for the loophole I
thought of with the `-w` option). Unfortunately the other things took more time

soup/location.c
/*
 * location - location/country codes
 *
 * "Because there is an I in IOCCC." :-)
 *
 * Copyright (c) 2022 by Landon Curt Noll.  All Rights Reserved.
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright, this permission notice and text
 * this comment, and the disclaimer below appear in all of the following:
 *
 *       supporting documentation
 *       source copies
 *       source works derived from this source
 *       binaries derived from this source or from derived source
 *
 * LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO
 * EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
 * USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 *
 * chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
 *
 * Share and enjoy! :-)
 */


/* special comments for the seqcexit tool */
/* exit code out of numerical order - ignore in sequencing - ooo */
/* exit code change of order - use new value in sequencing - coo */


#include <stdio.h>
#include <strings.h> /* strcasecmp */

/*
 * location - location/country codes
 */
#include "location.h"


/*
 * location/country codes
 *
 * We list all Officially assigned code elements,
 * all Exceptionally reserved code elements,
 * all Transitionally reserved code elements,
 * all Indeterminately reserved code elements,
 * all Formerly assigned code elements,
 * all User-assigned code elements,
 * as of 2021 Dec 31.  See:
 *
 *      https://en.wikipedia.org/wiki/ISO_3166-1#Officially_assigned_code_elements
 *      https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
 *      https://www.iso.org/obp/ui/#iso:pub:PUB500001:en
 *      https://www.iso.org/obp/ui/#search
 *
 * See also:
 *
 *      https://www.iso.org/glossary-for-iso-3166.html
 *
 * This means you will find a few codes that belong to
 * entities that no longer are recognized, entities
 * that are/were temporary, codes that are reserved for
 * future use, codes for Road Traffic Conventions,
 * codes for Property Organizations, duplicate codes, etc.
 *
 * We do not list WIPO codes as they are not formally
 * reserved.  They refer to Patent and related WIPO
 * organizations: as such they do not represent places
 * where an IOCCC winner would live.
 *
 * We mean no offense by this list: we simply tried to
 * include all ISO 3166 codes.  Please pardon any typos.
 * Spelling corrections welcome.
 *
 */
struct location loc[] = {
    {"AA", "User-assigned code AA"},	/* User-assigned code */
    {"AC", "Ascension Island"},	/* Exceptionally reserved code */
    {"AD", "Andorra"},
    {"AE", "United Arab Emirates"},
    {"AF", "Afghanistan"},
    {"AG", "Antigua and Barbuda"},
    {"AI", "Anguilla"},
    {"AL", "Albania"},
    {"AM", "Armenia"},
    {"AN", "Netherlands Antilles"},	/* Transitionally reserved code */
    {"AO", "Angola"},
/*  {"AP", "African Regional Industrial Property Organization"}, */ /* WIPO Indeterminately reserved code */
    {"AQ", "Antarctica"},
    {"AR", "Argentina"},
    {"AS", "American Samoa"},
    {"AT", "Austria"},
    {"AU", "Australia"},
    {"AW", "Aruba"},
    {"AX", "\xc3\x85land Islands"},
    {"AZ", "Azerbaijan"},
    {"BA", "Bosnia and Herzegovina"},
    {"BB", "Barbados"},
    {"BD", "Bangladesh"},
    {"BE", "Belgium"},
    {"BF", "Burkina Faso"},
    {"BG", "Bulgaria"},
    {"BH", "Bahrain"},
    {"BI", "Burundi"},
    {"BJ", "Benin"},
    {"BL", "Saint Barth\xc3\xa9lemy"},
    {"BM", "Bermuda"},
    {"BN", "Brunei Darussalam"},
    {"BO", "Bolivia (Plurinational State of)"},
    {"BQ", "Bonaire, Sint Eustatius and Saba"},
    {"BR", "Brazil"},
    {"BS", "Bahamas"},
    {"BT", "Bhutan"},
    {"BU", "Burma"},		/* Transitionally reserved code */
    {"BV", "Bouvet Island"},
    {"BW", "Botswana"},
/*  {"BX", "Benelux Trademarks and Designs Office"}, */ /* WIPO Indeterminately reserved code */
    {"BY", "Belarus"},
    {"BZ", "Belize"},
    {"CA", "Canada"},
    {"CC", "Cocos (Keeling) Islands (the)"},
    {"CD", "Congo, Democratic Republic of the"},
    {"CF", "Central African Republic"},
    {"CG", "Congo (the)"},
    {"CH", "Switzerland"},
    {"CI", "C\xc3\xb4te d'Ivoire"},
    {"CK", "Cook Islands"},
    {"CL", "Chile"},
    {"CM", "Cameroon"},
    {"CN", "China"},
    {"CO", "Colombia"},
    {"CP", "Clipperton Island"},	/* Exceptionally reserved code */
    {"CQ", "island of Sark"},	/* Exceptionally reserved code */
    {"CR", "Costa Rica"},
    {"CS", "Serbia and Montenegro"},	/* Transitionally reserved code */
    {"CT", "Canton and Enderbury Island"},	/* Formerly assigned code */
    {"CU", "Cuba"},
    {"CV", "Cabo Verde"},
    {"CW", "Cura\xc3\xa7" "ao"},
    {"CX", "Christmas Island"},
    {"CY", "Cyprus"},
    {"CZ", "Czechia"},
    {"DD", "German Democratic Republic"},	/* Formerly assigned code */
    {"DE", "Germany"},
    {"DG", "Diego Garcia"},	/* Exceptionally reserved code */
    {"DJ", "Djibouti"},
    {"DK", "Denmark"},
    {"DM", "Dominica"},
    {"DO", "Dominican Republic"},
    {"DY", "Benin"},		/* Indeterminately reserved code */
    {"DZ", "Algeria"},
    {"EA", "Ceuta, Melilla"},	/* Exceptionally reserved code */
    {"EC", "Ecuador"},
    {"EE", "Estonia"},
/*  {"EF", "Union of Countries under the European Community Patent Convention"}, */ /* WIPO Indeterminately reserved code */
    {"EG", "Egypt"},
    {"EH", "Western Sahara"},
/*  {"EM", "European Trademark Office"}, */ /* WIPO Indeterminately reserved code */
/*  {"EP", "European Patent Organization"}, */ /* WIPO Indeterminately reserved code */
    {"ER", "Eritrea"},
    {"ES", "Spain"},
    {"ET", "Ethiopia"},
    {"EU", "European Union"},	/* Exceptionally reserved code */
/*  {"EV", "Eurasian Patent Organization"}, */ /* WIPO Indeterminately reserved code */
    {"EW", "Estonia"},		/* Indeterminately reserved code */
    {"EZ", "European OTC derivatives"},	/* Exceptionally reserved code */
    {"FI", "Finland"},
    {"FJ", "Fiji"},
    {"FK", "Falkland Islands (the) [Malvinas]"},
    {"FL", "Liechtenstein"},	/* Indeterminately reserved code */
    {"FM", "Micronesia (Federated States of)"},
    {"FO", "Faroe Islands"},
    {"FQ", "French Southern and Antarctic Territories"},	/* Formerly assigned code */
    {"FR", "France"},
    {"FX", "France, Metropolitan"},	/* Exceptionally reserved code */
    {"GA", "Gabon"},
    {"GB", "United Kingdom of Great Britain and Northern Ireland (the)"},
/*  {"GC", "Patent Office of the Cooperation Council for the Arab States of the Gulf"}, */ /* WIPO Indeterminately reserved code */
    {"GD", "Grenada"},
    {"GE", "Georgia"},
    {"GF", "French Guiana"},
    {"GG", "Guernsey"},
    {"GH", "Ghana"},
    {"GI", "Gibraltar"},
    {"GL", "Greenland"},
    {"GM", "Gambia"},
    {"GN", "Guinea"},
    {"GP", "Guadeloupe"},
    {"GQ", "Equatorial Guinea"},
    {"GR", "Greece"},
    {"GS", "South Georgia and the South Sandwich Islands"},
    {"GT", "Guatemala"},
    {"GU", "Guam"},
    {"GW", "Guinea-Bissau"},
    {"GY", "Guyana"},
    {"HK", "Hong Kong"},
    {"HM", "Heard Island and McDonald Islands"},
    {"HN", "Honduras"},
    {"HR", "Croatia"},
    {"HT", "Haiti"},
    {"HU", "Hungary"},
    {"HV", "Upper Volta"},	/* Formerly assigned code */
/*  {"IB", "International Bureau of WIPO"}, */ /* WIPO Indeterminately reserved code */
    {"IC", "Canary Islands"},	/* Exceptionally reserved code */
    {"ID", "Indonesia"},
    {"IE", "Ireland"},
    {"IL", "Israel"},
    {"IM", "Isle of Man"},
    {"IN", "India"},
    {"IO", "British Indian Ocean Territory"},
    {"IQ", "Iraq"},
    {"IR", "Iran (Islamic Republic of)"},
    {"IS", "Iceland"},
    {"IT", "Italy"},
    {"JA", "Jamaica"},		/* Indeterminately reserved code */
    {"JE", "Jersey"},
    {"JM", "Jamaica"},
    {"JO", "Jordan"},
    {"JP", "Japan"},
    {"JT", "Johnston Island"},	/* Formerly assigned code */
    {"KE", "Kenya"},
    {"KG", "Kyrgyzstan"},
    {"KH", "Cambodia"},
    {"KI", "Kiribati"},
    {"KM", "Comoros"},
    {"KN", "Saint Kitts and Nevis"},
    {"KP", "Korea (the Democratic People's Republic of)"},
    {"KR", "Korea (the Republic of)"},
    {"KW", "Kuwait"},
    {"KY", "Cayman Islands"},
    {"KZ", "Kazakhstan"},
    {"LA", "Lao People's Democratic Republic"},
    {"LB", "Lebanon"},
    {"LC", "Saint Lucia"},
    {"LF", "Libya Fezzan"},	/* Indeterminately reserved code */
    {"LI", "Liechtenstein"},
    {"LK", "Sri Lanka"},
    {"LR", "Liberia"},
    {"LS", "Lesotho"},
    {"LT", "Lithuania"},
    {"LU", "Luxembourg"},
    {"LV", "Latvia"},
    {"LY", "Libya"},
    {"MA", "Morocco"},
    {"MC", "Monaco"},
    {"MD", "Moldova (the Republic of)"},
    {"ME", "Montenegro"},
    {"MF", "Saint Martin (French part)"},
    {"MG", "Madagascar"},
    {"MH", "Marshall Islands (the)"},
    {"MI", "Midway Islands"},	/* Formerly assigned code */
    {"MK", "North Macedonia"},
    {"ML", "Mali"},
    {"MM", "Myanmar"},
    {"MN", "Mongolia"},
    {"MO", "Macao"},
    {"MP", "Northern Mariana Islands"},
    {"MQ", "Martinique"},
    {"MR", "Mauritania"},
    {"MS", "Montserrat"},
    {"MT", "Malta"},
    {"MU", "Mauritius"},
    {"MV", "Maldives"},
    {"MW", "Malawi"},
    {"MX", "Mexico"},
    {"MY", "Malaysia"},
    {"MZ", "Mozambique"},
    {"NA", "Namibia"},
    {"NC", "New Caledonia"},
    {"NE", "Niger"},
    {"NF", "Norfolk Island"},
    {"NG", "Nigeria"},
    {"NH", "New Hebrides"},	/* Formerly assigned code */
    {"NI", "Nicaragua"},
    {"NL", "Netherlands"},
    {"NO", "Norway"},
    {"NP", "Nepal"},
    {"NQ", "Dronning Maud Land"},	/* Formerly assigned code */
    {"NR", "Nauru"},
    {"NT", "Neutral Zone"},	/* Transitionally reserved code */
    {"NU", "Niue"},
    {"NZ", "New Zealand"},
/*  {"OA", "African Intellectual Property Organization"}, */ /* WIPO Indeterminately reserved code */
    {"OO", "user-assigned escape code"},	/* User-assigned code */
    {"OM", "Oman"},
    {"PA", "Panama"},
    {"PC", "Pacific Islands Trust Territory"},	/* Formerly assigned code */
    {"PE", "Peru"},
    {"PF", "French Polynesia"},
    {"PG", "Papua New Guinea"},
    {"PH", "Philippines"},
    {"PI", "Philippines"},	/* Indeterminately reserved code */
    {"PK", "Pakistan"},
    {"PL", "Poland"},
    {"PM", "Saint Pierre and Miquelon"},
    {"PN", "Pitcairn"},
    {"PR", "Puerto Rico"},
    {"PS", "Palestine, State of"},
    {"PT", "Portugal"},
    {"PU", "United States Miscellaneous Pacific Islands"},	/* Formerly assigned code */
    {"PW", "Palau"},
    {"PY", "Paraguay"},
    {"PZ", "Panama Canal Zone"},	/* Formerly assigned code */
    {"QA", "Qatar"},
    {"QM", "User-assigned code QM"},	/* User-assigned code */
    {"QN", "User-assigned code QN"},	/* User-assigned code */
    {"QO", "User-assigned code QO"},	/* User-assigned code */
    {"QP", "User-assigned code QP"},	/* User-assigned code */
    {"QQ", "User-assigned code QQ"},	/* User-assigned code */
    {"QR", "User-assigned code QR"},	/* User-assigned code */
    {"QS", "User-assigned code QS"},	/* User-assigned code */
    {"QT", "User-assigned code QT"},	/* User-assigned code */
    {"QU", "User-assigned code QU"},	/* User-assigned code */
    {"QV", "User-assigned code QV"},	/* User-assigned code */
    {"QW", "User-assigned code QW"},	/* User-assigned code */
    {"QX", "User-assigned code QX"},	/* User-assigned code */
    {"QY", "User-assigned code QY"},	/* User-assigned code */
    {"QZ", "User-assigned code QZ"},	/* User-assigned code */
    {"RA", "Argentina"},	/* Indeterminately reserved code */
    {"RB", "Bolivia or Botswana"},	/* Indeterminately reserved code */
    {"RC", "China"},		/* Indeterminately reserved code */
    {"RE", "R\xc3\xa9union"},
    {"RH", "Haiti"},		/* Indeterminately reserved code */
    {"RI", "Indonesia"},	/* Indeterminately reserved code */
    {"RL", "Lebanon"},		/* Indeterminately reserved code */
    {"RM", "Madagascar"},	/* Indeterminately reserved code */
    {"RN", "Niger"},		/* Indeterminately reserved code */
    {"RO", "Romania"},
    {"RP", "Philippines"},	/* Indeterminately reserved code */
    {"RS", "Serbia"},
    {"RU", "Russian Federation"},
    {"RW", "Rwanda"},
    {"SA", "Saudi Arabia"},
    {"SB", "Solomon Islands"},
    {"SC", "Seychelles"},
    {"SD", "Sudan"},
    {"SE", "Sweden"},
    {"SF", "Finland"},		/* Indeterminately reserved code */
    {"SG", "Singapore"},
    {"SH", "Saint Helena, Ascension and Tristan da Cunha"},
    {"SI", "Slovenia"},
    {"SJ", "Svalbard and Jan Mayen"},
    {"SK", "Slovakia"},
    {"SL", "Sierra Leone"},
    {"SM", "San Marino"},
    {"SN", "Senegal"},
    {"SO", "Somalia"},
    {"SR", "Suriname"},
    {"SS", "South Sudan"},
    {"ST", "Sao Tome and Principe"},
    {"SU", "USSR"},		/* Exceptionally reserved code */
    {"SV", "El Salvador"},
    {"SX", "Sint Maarten (Dutch part)"},
    {"SY", "Syrian Arab Republic"},
    {"SZ", "Eswatini"},
    {"TA", "Tristan da Cunha"},	/* Exceptionally reserved code */
    {"TC", "Turks and Caicos Islands"},
    {"TD", "Chad"},
    {"TF", "French Southern Territories"},
    {"TG", "Togo"},
    {"TH", "Thailand"},
    {"TJ", "Tajikistan"},
    {"TK", "Tokelau"},
    {"TL", "Timor-Leste"},
    {"TM", "Turkmenistan"},
    {"TN", "Tunisia"},
    {"TO", "Tonga"},
    {"TP", "Turkey"},		/* Transitionally reserved code */
    {"TR", "Turkey"},
    {"TT", "Trinidad and Tobago"},
    {"TV", "Tuvalu"},
    {"TW", "Taiwan"},
    {"TZ", "Tanzania, United Republic of"},
    {"UA", "Ukraine"},
    {"UG", "Uganda"},
    {"UK", "United Kingdom"},	/* Exceptionally reserved code */
    {"UM", "United States Minor Outlying Islands"},
    {"UN", "United Nations"},	/* Exceptionally reserved code */
    {"US", "United States of America"},
    {"UY", "Uruguay"},
    {"UZ", "Uzbekistan"},
    {"VA", "Holy See"},
    {"VC", "Saint Vincent and the Grenadines"},
    {"VD", "Democratic Republic of Viet-Nam"},	/* Formerly assigned code */
    {"VE", "Venezuela (Bolivarian Republic of)"},
    {"VG", "Virgin Islands (British)"},
    {"VI", "Virgin Islands (U.S.)"},
    {"VN", "Viet Nam"},
    {"VU", "Vanuatu"},
    {"WF", "Wallis and Futuna"},
    {"WG", "Grenada"},		/* Indeterminately reserved code */
    {"WK", "Wake Island"},	/* Formerly assigned code */
    {"WL", "Saint Lucia"},	/* Indeterminately reserved code */
/*  {"WO", "World Intellectual Property Organization"}, */ /* WIPO Indeterminately reserved code */
    {"WS", "Samoa"},
    {"WV", "Saint Vincent"},	/* Indeterminately reserved code */
    {"XA", "User-assigned code XA"},	/* User-assigned code */
    {"XB", "User-assigned code XB"},	/* User-assigned code */
    {"XC", "User-assigned code XC"},	/* User-assigned code */
    {"XD", "User-assigned code XD"},	/* User-assigned code */
    {"XE", "User-assigned code XE"},	/* User-assigned code */
    {"XF", "User-assigned code XF"},	/* User-assigned code */
    {"XG", "User-assigned code XG"},	/* User-assigned code */
    {"XH", "User-assigned code XH"},	/* User-assigned code */
    {"XI", "User-assigned code XI"},	/* User-assigned code */
    {"XJ", "User-assigned code XJ"},	/* User-assigned code */
    {"XK", "User-assigned code XK"},	/* User-assigned code */
    {"XL", "User-assigned code XL"},	/* User-assigned code */
    {"XM", "User-assigned code XM"},	/* User-assigned code */
    {"XN", "User-assigned code XN"},	/* User-assigned code */
    {"XO", "User-assigned code XO"},	/* User-assigned code */
    {"XP", "User-assigned code XP"},	/* User-assigned code */
    {"XR", "User-assigned code XR"},	/* User-assigned code */
    {"XS", "User-assigned code XS"},	/* User-assigned code */
    {"XT", "User-assigned code XT"},	/* User-assigned code */
    {"XU", "User-assigned code XU"},	/* User-assigned code */
    {"XV", "User-assigned code XV"},	/* User-assigned code */
    {"XW", "User-assigned code XW"},	/* User-assigned code */
    {"XX", "User-assigned code XX for anonymous locations"},	/* User-assigned code */
    {"XY", "User-assigned code XY"},	/* User-assigned code */
    {"XZ", "User-assigned code XZ"},	/* User-assigned code */
    {"YD", "Democratic Yemen"},	/* Formerly assigned code */
    {"YE", "Yemen"},
    {"YT", "Mayotte"},
    {"YU", "Yugoslavia"},	/* Transitionally reserved code */
    {"YV", "Venezuela"},	/* Indeterminately reserved code */
    {"ZA", "South Africa"},
    {"ZM", "Zambia"},
    {"ZR", "Zaire"},		/* Transitionally reserved code */
    {"ZW", "Zimbabwe"},
    {"ZZ", "User-assigned code ZZ"},	/* User-assigned code */
    {NULL, NULL}		/* MUST BE LAST */
};

size_t SIZEOF_LOCATION_TABLE = TBLLEN(loc);


/* check_location_table	    - make sure that there are no embedded NULL elements
 *			      in the location table and that the last element is
 *			      NULL
 *
 * This function verifies that the only NULL element in the table is the very
 * last element: if it's not there or there's another NULL element it's a
 * problem that has to be fixed.
 *
 * This function does not return on error.
 */
void
check_location_table(void)
{
    size_t max = SIZEOF_LOCATION_TABLE;

    size_t i;

    for (i = 0; i < max - 1 && loc[i].code != NULL; )
	++i;

    if (max - 1 != i) {
	err(333, __func__, "found embedded NULL element in location table; fix table in %s and recompile", __FILE__);
	not_reached();
    }
    if (loc[i].code != NULL || loc[i].name != NULL) {
	err(10, __func__, "no final NULL element found in location table; fix table in %s and recompile", __FILE__);
	not_reached();
    }
}


/*
 * lookup_location_name - convert a ISO 3166-1 Alpha-2 into a location name
 *
 * Even though the ISO 3166-1 Alpha-2 code is only UPPER CASE, the
 * code is compared in a case independent way.  That is, a non-canonical
 * lower case code will match the canonical UPPER CASE ISO 3166-1 Alpha-2 code.
 *
 * given:
 *      code      ISO 3166-1 Alpha-2 code
 *
 * return:
 *      location name or NULL ==> unlisted code
 *
 * This function does not return on error.
 */
char const *
lookup_location_name(char const *code)
{
    struct location *p;		/* entry in the location table */
    size_t length = 0;		/* length of code */

    /*
     * firewall
     */
    if (code == NULL) {
	err(11, __func__, "code is NULL");
	not_reached();
    }

    /*
     * sanity check - code must be 2 characters long
     */
    length = strlen(code);
    if (length != 2) {
	dbg(DBG_HIGH, "code: <%s> length: %zu != 2", code, length);
	return NULL;
    }

    /*
     * sanity check - code must be 2 ASCII alpha characters
     */
    if (!isascii(code[0]) || !isalpha(code[0]) ||
	!isascii(code[1]) || !isalpha(code[1])) {
	dbg(DBG_HIGH, "code: <%s> is not 2 ASCII alpha characters", code);
	return NULL;
    }

    /*
     * search location table for the code
     */
    for (p = &loc[0]; p->code != NULL && p->name != NULL; ++p) {
	if (strcasecmp(code, p->code) == 0) {
	    dbg(DBG_VHIGH, "code: %s found name: <%s>", p->code, p->name);
	    break;
	}
    }

    /*
     * return name or NULL
     */
    if (p->name == NULL) {
	dbg(DBG_HIGH, "code: <%s> is unknown", code);
    }
    return p->name;
}


/*
 * lookup_location_code - convert a ISO 3166-1 Alpha-2 into a location name
 *
 * Even though the location table contains a canonical name in a canonical case,
 * we compare names in a case independent way.  Nevertheless, the ISO 3166-1 Alpha-2
 * code returned will be a 2 character ASCII UPPER CASE code.
 *
 * given:
 *      location_name      location name of a ISO 3166-1 Alpha-2 entry
 *
 * return:
 *	ISO 3166-1 Alpha-2 in UPPER CASE code or NULL ==> unknown location name
 *
 * This function does not return on error.
 */
char const *
lookup_location_code(char const *location_name)
{
    struct location *p;		/* entry in the location table */

    /*
     * firewall
     */
    if (location_name == NULL) {
	err(12, __func__, "location_name is NULL");
	not_reached();
    }

    /*
     * search location table for the code
     */
    for (p = &loc[0]; p->name != NULL && p->code != NULL; ++p) {
	if (strcasecmp(location_name, p->name) == 0) {
	    dbg(DBG_VHIGH, "name: <%s> found code: %s", p->name, p->code);
	    break;
	}
    }

    /*
     * return code or NULL
     */
    if (p->code == NULL) {
	dbg(DBG_HIGH, "name: <%s> is unknown", location_name);
    }
    return p->code;
}


/*
 * location_code_name_match - if a location code & location name refer to the same place
 *
 * We look up a ISO 3166-1 Alpha-2 in UPPER CASE code and determine if the associated
 * location name matches the given location name.
 *
 * given:
 *	code	ISO 3166-1 Alpha-2 in UPPER CASE
 *	location_name	location name of a ISO 3166-1 Alpha-2 entry
 *
 * return:
 *	true ==> location code and location name match,
 *	false ==> location code and location name do not match or,
 *		  unknown location code, or NULL ptr
 */
bool
location_code_name_match(char const *code, char const *location_name)
{
    struct location *p;		/* entry in the location table */

    /*
     * firewall
     */
    if (code == NULL) {
	err(13, __func__, "code is NULL");
	not_reached();
    }
    if (location_name == NULL) {
	err(14, __func__, "location_name is NULL");
	not_reached();
    }

    /*
     * search location table for the code
     */
    for (p = &loc[0]; p->code != NULL && p->name != NULL; ++p) {
	if (strcasecmp(code, p->code) == 0) {
	    dbg(DBG_VHIGH, "code: %s found name: <%s>", p->code, p->name);
	    break;
	}
    }
    if (p->code == NULL) {
	dbg(DBG_HIGH, "code: <%s> is unknown", code);
	return false;
    }

    if (p->name == NULL) {
	dbg(DBG_HIGH, "name: <%s> is unknown", location_name);
	return false;
    }

    /*
     * compare location name with name from the code found in the table
     */
    if (strcasecmp(location_name, p->name) != 0) {
	dbg(DBG_HIGH, "code: %s does not match name: <%s>", code, location_name);
	return false;
    }

    /*
     * return match
     */
    return true;
}
